doc/design/persistent_volumes_etcd_data.md

```diff
@@ -0,0 +1,125 @@
+* k8s nodes scheduling a majority of etcd members pods reboots and come back before node controller pod eviction triggers
+
+##### Effect
+
+The pods will go in failed state, the etcd cluster becomes unquorate. Etcd-operator can only restore the cluster from a backup.
+
+#### Case D
+
+##### Event
+
+* k8s nodes executing a majority of etcd members pods becomes partitioned (network partitioning, goes down, etc...) and cannot talk to the api server, after some time the node controller will evict their pods.
+
+##### Effect
+
+The pods will go in an unknown/failed state and eventually be removed from the API server, the etcd cluster becomes unquorate. Etcd-operator can only restore the cluster from a backup.
+
+
+#### Reasons
+
+Currently etcd operator handles cases A and B but when a cluster becomes unquorate (cases C and D) the current option is to restore the cluster from a backup.
+
+User may desire consistency over availability and prefer to wait for the cluster the return in a quorate state instead of restoring from a backup (that will contain old data).
+
+This proposal aims to fix cases C and D permitting to the cluster the return in a quorate state without the need to restore from a backup.
+
+## Proposed Changes
+
+#### Changes
+
+The change to be effective and cover the above failure scenario is split in two parts.
+
+### Part 1
+
+- Add to the cluster spec PodPolicy an option to enable putting etcd data inside a Persistent Volume
+- If podPolicy has persistent volumes enabled, when creating a new member also create a Persistent Volume Claim and use it as the volume source for etcd-data
+- Handle etcd data persistent volume claims cleanup:
+  - set on the PVC an owner like already done with other resources.
+  - in the reconcile loop, remove PVCs that don't have a matching etcd member in the etcd cluster and their related pod doesn't exists in the k8s API
+  - garbage collect also PVCs
+
+#### Pod name to Persistent Volume Claim name mapping
+
+The pod and related PVC are mapped 1 to 1 so their names be easily generated and deducted. I.E. if the pod is called `etcd-0000` its related PVC name will be `etcd-0000-pvc`
+
+#### Results
+
+These changes will put the basis for initial persistent volume management
+
+#### Problems
+
+This first part won't fix case C and D.
+
+### Part 2
+
+#### Changes
+
+- Change reconcile logic to handle deleted pods that had a persistent volume:
+
+If a pod of an etcd cluster member is deleted and it had a persistent volume claim associated, instead of removing the etcd member from the etcd cluster and adding a new member just recreate it (keeping its pod name) binding the existing persistent volume claim. The pod recreation will happen only if the pod doesn't exists anymore in the API because we'll keep the same pod name. This also enable a MANDATORY property: at most once pod existence logic to avoid having in the API at the same time two pods with associated the same PVC.
```

Comment:
So this paragraph only assumes one failure scenario -- pod is deleted.

See comment https://github.com/coreos/etcd-operator/pull/1374/files#r136628955 .

Please write down the two failure scenarios in part 2.

Reply:
Sorry but I lost what other failure scenario you're talking about.

Reply:
Sure. Let me clarify what should be done in part 2 more.


### Problems

What problems we have left in part 1? Failures:

- pod failure: pod deleted (for some reason), pod SLA low (better recreate it somewhere else)
- volume failure (failing to bind): In the future, k8s will add a taint to PV and kubelet will evict the pod. Pod unready -> pod deleted.
- node failure (eviction): pod unready -> pod deleted.
- scheduling failure: pod pending

So you are mostly right that pod being deleted is the most possible failures. But there could be other situations that we want to delete the pod in operator.

### Solutions

For pod deleted, we could just recreate the pod.

However, we also need to have tolerations. For example, we should set tolerations on volume taint, node taint. And we should tolerate pod "unready" for whatever reason for maximum time.

After a second thought, I think we could either deal with both recreating the pod and tolerations in part 2, or separate tolerations into part 3, or just leave tolerations as TODO (we can keep on working on it later).

Reply:
Thanks for the explanation, I agree with you. Perhaps a part 3 approach will be a good step but I fear that until it's implemented, the part 2 implementation won't gracefully handle some failures (i.e. corrupted data) that are today handled with the current ephemeral pods. What's your preference?

Reply:
No preference. All is good.